app/src/main/java/org/stepik/android/domain/course_purchase/interactor/CoursePurchaseInteractor.kt

```diff
@@ -23,8 +65,81 @@ constructor(
                     Single.just(PromoCodeSku.EMPTY)
                 } else {
                     lightSkuRepository
-                        .getLightInventory(ProductTypes.IN_APP, listOf(mobileTier.promoTier), dataSourceType = DataSourceType.REMOTE)
+                        .getLightInventory(BillingClient.SkuType.INAPP, listOf(mobileTier.promoTier), dataSourceType = DataSourceType.REMOTE)
                         .map { lightSku -> PromoCodeSku(promoCodeName, lightSku.firstOrNull()) }
                 }
             }
+
+    fun launchPurchaseFlow(courseId: Long, skuId: String): Single<Pair<CoursePurchaseObfuscatedParams, SkuDetails>> =
+        coursePaymentsRepository
+            .getCoursePaymentsByCourseId(courseId, CoursePayment.Status.SUCCESS, sourceType = DataSourceType.REMOTE)
+            .flatMap { payments ->
+                if (payments.isEmpty()) {
+                    getSkuDetails(courseId, skuId)
+                } else {
+                    Single.error(CourseAlreadyOwnedException(courseId))
+                }
+            }
+
+    fun completePurchase(courseId: Long, sku: SkuDetails, purchase: Purchase): Completable =
+        coursePaymentsRepository
+            .createCoursePayment(courseId, sku, purchase)
+            .flatMapCompletable { payment ->
+                if (payment.status == CoursePayment.Status.SUCCESS) {
+                    Completable.complete()
+                } else {
+                    Completable.error(CoursePurchaseVerificationException())
+                }
+            }
+            .andThen(updateCourseAfterEnrollment(courseId))
+            .andThen(billingRepository.consumePurchase(purchase))
+
+    fun restorePurchase(courseId: Long, skuId: String): Completable =
```

Comment:
Will it work if user bought course with promo code in past? Or course price was changed.